### ISSUE TITLE
oracle support

### ISSUE THREAD

**Issue by codyzu** [pouchdb/pouchdb#5283]
### Environment (Node.js/browser/hybrid app/etc.)

node.js

### Browser/platform (Chrome/FF/Safari/Edge/iOS/Android/etc.)

N/A

### Adapter (IndexedDB/WebSQL/LevelDB/etc.)

LevelDown -> SQLDown

### Server (CouchDB/Cloudant/Couchbase/PouchDB Server/etc.)

Oracle 12c

### Please describe your issue. Test cases (e.g. JSBin) are very helpful!

During a spike for my project, I have put together a proof-of-concept using PouchDb with Oracle. My POC works with: PouchDB -> LevelDB Adapter -> [SQLDown](https://github.com/calvinmetcalf/SQLdown) -> [Knex](https://github.com/tgriesser/knex) -> [oracledb node.js driver](https://github.com/oracle/node-oracledb) -> Oracle.

There are several caveats:

1. Knex does not support the [official oracle node.js driver](https://github.com/oracle/node-oracledb). There is a PR tgriesser/knex#990 to add support for the oracledb driver (which I used), but I found at least 1 issue with the stream functionality.
2. SQLDown, while internally using knex, in practice is written to support websql (in browser) and sqlite3 or mysql in the node.js. This resulted in some problems with the table schema that I had to hack around.

In the end, I found that this implementation was generally half as performant as using the oracledb node.js driver directly (or via knex, using the merge request mentioned above).

I would like to return some of this work back to the community and ultimately open the door to PouchDB working with Oracle.

Before I do, I have some high level questions:

1. If PouchDB supports Oracle, would it be better using the LevelDB Adapter -> SQLDown *or* implementing a custom adapter (maybe via a PouchDb plugin) to interact directly to the oracledb driver (or knex), similar to the websql adapter? Does the SQLDown adapter simply create too much complexity to ever be performant?
2. Is it completely crazy to think that PouchDB would ever be "performant" when using SQL (oracle) backend?
3. Is this a good idea? :-)

Thanks for the great project!

**Comment by daleharvey** [comment]
1. So @nolanlawson will probably be able to answer this better than me, however I believe the ideal adapter to base it on would be 'pouch-adapter-websql', this is a direct translation between pouchdb calls to sql and I am assuming that there wont be too much of a difference between what oracle supports and sqlite (although likely some work needed), basically the less levels of indirections the better. We have 'pouch-adapter-node-websql' to allow this to work in node which might be used as an example of how to get it working against oracle. 

2. So there is going to be some cost with pouchdb managing revisions, however I dont think it should be a particulary large difference and I can see pouchdb being performance. If the current adapters are too slow I am confident that a websql adapter based on the `idb-next` format (https://github.com/pouchdb/pouchdb/tree/master/packages/pouchdb-adapter-indexeddb) will be plenty fast.

3. Thats the main question for me, why use oracle as a backend? if there are ops / departmental reasons that oracle is an absolute requirement then I believe it would be possible to make it work, however given the choice I would always go with the supported versions, ie couchdb or the leveldb adapter for the server. Remember that you will not be able to sync existing oracle data or access pouchdb data via the oracle api (https://pouchdb.com/faq.html#sync_non_couchdb)

**Comment by daleharvey** [comment]
Closing out since its not work we will be tracking in pouch, but feel free to carry on the discussion

**Comment by nolanlawson** [comment]
Yep, you could write a custom adapter ala `pouchdb-adapter-node-websql` and using the [node-websql custom API](https://github.com/nolanlawson/node-websql#custom-sqlite3-bindings). However I doubt very much it would work since the adapter uses SQLite and not generic SQL, e.g. it even reads from the `sqlite_master` database which does not exist in Oracle.

**Comment by nolanlawson** [comment]
And in general: no, I don't think this is a good idea. I think most people who try to get MySQL/Postgres/Oracle/etc. working with PouchDB _actually_ just want an easy and familiar way to query some database using SQL (since CouchDB views are so complicated by comparison). For that, you can just write a cron job to periodically dump CouchDB to a SQL database in whatever format you want, and then you're done.

**Comment by codyzu** [comment]
@daleharvey @nolanlawson Thanks for the great discussion!

A little note about the motivation... Most of our clients are large enterprises, typically in the financial and government sectors, and they haven't given us much choice about the database (since Oracle is well established in large enterprises). Our choice is Oracle or Oracle. A bit to my disappointment, the oracle support in node seems a little week. That is why I attempted to abstract up to PouchDb.  I tried to give a nice javascript API of PouchDb with Oracle on the backend.

I think in the near term we will use knex directly. The cost to get PouchDb -> Oracle working reliably seems a high. Especially considering we don't really need any of the cool syncing or conflict resolution features.

I still plan to push my changes upstream to knex and sqldown... with the goal to at least ensure there is a working solution if anyone else is interested in the this crazy dream of an idea :-)
